<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>isoRelate Introduction • isoRelate</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><meta property="og:title" content="isoRelate Introduction">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">isoRelate</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/introduction.html">isoRelate Introduction</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>isoRelate Introduction</h1>
                        <h4 class="author">Lyndal Henden</h4>
            
            <h4 class="date">2018-03-15</h4>
          </div>

    
    
<div class="contents">
<p>isoRelate performs pairwise relatedness mapping on haploid isolates using a first order continuous time hidden Markov model (HMM). Some isolates may have multiple infections, whereby the infected individual is carrying multiple genetically distinct strains of the species. Such isolates are said to have multiplicity of infection (MOI) greater than 1, and isoRelate allows for these isolates by analyzing them using a diploid model, rather than a haploid model.</p>
<p>This document outlines the data format and key functions used by isoRelate for relatedness mapping.</p>
<p><strong>Caution should be taken when using isoRelate with small reference datasets, reference populations that do not match the input population or cohorts of mixed populations.</strong></p>
<div id="data-format" class="section level2">
<h2 class="hasAnchor">
<a href="#data-format" class="anchor"></a>Data Format</h2>
<div id="ped-and-map-files" class="section level3">
<h3 class="hasAnchor">
<a href="#ped-and-map-files" class="anchor"></a>PED and MAP Files</h3>
<p>The input for isoRelate is unphased genotype data for SNPs in PLINK PED and MAP format (<a href="http://www.cog-genomics.org/plink2" class="uri">http://www.cog-genomics.org/plink2</a>).</p>
<p>A PED file is a white-space (space or tab) delimited file with the first six columns</p>
<ol style="list-style-type: decimal">
<li>Family ID</li>
<li>Isolate ID</li>
<li>Paternal ID</li>
<li>Maternal ID</li>
<li>Multiplicity of infection (MOI) (1 = single infection or haploid, 2 = multiple infections or diploid)</li>
<li>Phenotype (1=unaffected, 2=affected, 0=unknown)</li>
</ol>
<p>The IDs are alphanumeric: the combination of family and isolate ID should uniquely identify a sample. Columns 7 onwards (white-space delimited) are the isolate genotypes for biallelic SNPs where the A and B alleles are coded as 1 and 2 respectively and missing genotypes are coded as 0. All SNPs must have two alleles specified and each allele should be in a separate column. For single infections, genotypes should be specified as homozygous. Either Both alleles should be missing (i.e. 0) or neither. Column labels are not required. The paternal ID, maternal ID and phenotype columns are not used by isoRelate, however are required for completeness of the pedigree. Examples of informative family IDs are the sample collection site or country, however family IDs can be the same as the isolate IDs.</p>
<p>The MAP file contains exactly 4 columns of information</p>
<ol style="list-style-type: decimal">
<li>Chromosome</li>
<li>SNP identifier</li>
<li>Genetic map distance (centi-Morgans or Morgans)</li>
<li>Base-pair position</li>
</ol>
<p>where each row describes a single marker. Genetic map distances and base-pair positions are expected to be positive values. The MAP file must be ordered by increasing chromosome genetic map distance. SNP identifiers can contain any characters except spaces or tabs; also, you should avoid * symbols in names. The MAP file must contain as many markers as are in the PED file. Column labels are not required.</p>
</div>
<div id="reference-data" class="section level3">
<h3 class="hasAnchor">
<a href="#reference-data" class="anchor"></a>Reference Data</h3>
<p>isoRelate calculates SNP allele frequencies from either the input dataset or a reference dataset, where the reference dataset is in PED/MAP format. If a reference dataset is not specified then allele frequencies will be calculated from the input dataset.</p>
<p><strong>Caution should be taken when using isoRelate with small reference datasets, reference populations that do not match the input population or cohorts of mixed populations.</strong></p>
</div>
</div>
<div id="key-functions" class="section level2">
<h2 class="hasAnchor">
<a href="#key-functions" class="anchor"></a>Key Functions</h2>
<table style="width:97%;" class="table">
<colgroup>
<col width="26%">
<col width="70%">
</colgroup>
<thead><tr class="header">
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>getGenotypes</td>
<td>reformat the genotypes in the PED/MAP file and perform data filtering</td>
</tr>
<tr class="even">
<td>getIBDparameters</td>
<td>estimate IBD parameters between pairs of isolates</td>
</tr>
<tr class="odd">
<td>getIBDsegments</td>
<td>detect IBD segments between pairs of isolates</td>
</tr>
<tr class="even">
<td>getIBDposterior</td>
<td>calculate the posterior probability of IBD sharing per isolate pair</td>
</tr>
<tr class="odd">
<td>getIBDsummary</td>
<td>print a summary of the IBD segments detected</td>
</tr>
<tr class="even">
<td>getIBDmatrix</td>
<td>create a binary matrix of IBD/non-IBD for all SNP and pairs</td>
</tr>
<tr class="odd">
<td>getIBDproportions</td>
<td>calculate the proportion of pairs IBD at each SNP</td>
</tr>
<tr class="even">
<td>getIBDiR</td>
<td>calculate the significance of excess IBD sharing at each SNP</td>
</tr>
<tr class="odd">
<td>getIBDiclusters</td>
<td>create a network of clusters of isolates who are IBD over an <strong>interval</strong>
</td>
</tr>
<tr class="even">
<td>getIBDpclusters</td>
<td>create a network of clusters of isolates who share a certain <strong>proportion</strong> of genome IBD</td>
</tr>
<tr class="odd">
<td>plotIBDsegments</td>
<td>plot the detected IBD segments across the genome</td>
</tr>
<tr class="even">
<td>plotIBDproportions</td>
<td>plot the proportion of pairs IBD across the genome</td>
</tr>
<tr class="odd">
<td>plotIBDiR</td>
<td>plot the significance of excess IBD sharing across the genome</td>
</tr>
<tr class="even">
<td>plotIBDclusters</td>
<td>plot the network of IBD isolates</td>
</tr>
</tbody>
</table>
</div>
<div id="example" class="section level2">
<h2 class="hasAnchor">
<a href="#example" class="anchor"></a>Example</h2>
<p>Below is an IBD analysis performed on a <em>Plasmodium falciparum</em> dataset from Papua New Guinea (PNG), generated by the MalariaGEN Consortium (<a href="https://www.malariagen.net/projects/pf3k" class="uri">https://www.malariagen.net/projects/pf3k</a>). Change the parameters for each function to see how it affects the results.</p>
<p>**NOTE: R has a memory limit and if you exceed this limit then an error message will appear stating <code>cannot allocate vector of length</code>. The number of bytes in a character string is limited to 2^31 - 1 ~ 2*10^9, which is also the limit on each dimension of an array.**</p>
<div id="format-and-filter-pedmap" class="section level3">
<h3 class="hasAnchor">
<a href="#format-and-filter-pedmap" class="anchor"></a>Format and Filter PED/MAP</h3>
<p>We begin by reformatting and filtering the input PED and MAP data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># load isoRelate library</span>
<span class="kw">library</span>(isoRelate)

<span class="co"># lets look at the data</span>
<span class="kw">str</span>(png_pedmap)</code></pre></div>
<pre><code>## List of 2
##  $ :'data.frame':    38 obs. of  59268 variables:
##   ..$ fid  : chr [1:38] "PNG" "PNG" "PNG" "PNG" ...
##   ..$ iid  : chr [1:38] "PN0003-C" "PN0004-C" "PN0005-C" "PN0006-C" ...
##   ..$ pid  : num [1:38] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ mid  : num [1:38] 0 0 0 0 0 0 0 0 0 0 ...
##   ..$ moi  : num [1:38] 1 1 1 1 2 1 1 2 1 1 ...
##   ..$ aff  : num [1:38] 2 2 2 2 2 2 2 2 2 2 ...
##   ..$ 1    : num [1:38] 1 1 1 0 1 1 1 1 1 1 ...
##   ..$ 2    : num [1:38] 1 1 1 0 1 1 1 1 1 1 ...
##   ..$ 3    : num [1:38] 0 1 1 0 1 1 1 1 1 1 ...
##   ..$ 4    : num [1:38] 0 1 1 0 1 1 1 1 1 1 ...
##   ..$ 5    : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 6    : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 7    : num [1:38] 1 1 1 1 1 1 2 1 1 1 ...
##   ..$ 8    : num [1:38] 1 1 1 1 2 1 2 2 1 1 ...
##   ..$ 9    : num [1:38] 1 1 1 2 1 1 1 1 1 1 ...
##   ..$ 10   : num [1:38] 1 1 1 2 1 1 1 1 1 1 ...
##   ..$ 11   : num [1:38] 1 1 0 2 1 1 1 1 1 1 ...
##   ..$ 12   : num [1:38] 1 1 0 2 1 1 1 1 1 1 ...
##   ..$ 13   : num [1:38] 1 1 0 1 1 1 1 1 2 1 ...
##   ..$ 14   : num [1:38] 1 1 0 1 1 1 1 1 2 1 ...
##   ..$ 15   : num [1:38] 1 1 0 1 1 1 1 1 1 0 ...
##   ..$ 16   : num [1:38] 1 1 0 1 1 1 1 1 1 0 ...
##   ..$ 17   : num [1:38] 1 1 0 1 1 1 1 1 1 1 ...
##   ..$ 18   : num [1:38] 1 1 0 1 1 1 1 1 1 1 ...
##   ..$ 19   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 20   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 21   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 22   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 23   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 24   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 25   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 26   : num [1:38] 1 1 1 1 2 1 1 1 1 1 ...
##   ..$ 27   : num [1:38] 2 2 2 1 1 2 1 1 1 1 ...
##   ..$ 28   : num [1:38] 2 2 2 1 1 2 1 1 1 1 ...
##   ..$ 29   : num [1:38] 2 2 2 1 1 2 1 1 1 1 ...
##   ..$ 30   : num [1:38] 2 2 2 1 1 2 1 1 1 1 ...
##   ..$ 31   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 32   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 33   : num [1:38] 1 1 1 1 1 1 1 1 2 1 ...
##   ..$ 34   : num [1:38] 1 1 1 1 1 1 1 1 2 1 ...
##   ..$ 35   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 36   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 37   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 38   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 39   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 40   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 41   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 42   : num [1:38] 1 1 1 1 2 1 1 1 1 1 ...
##   ..$ 43   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 44   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 45   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 46   : num [1:38] 1 1 1 1 1 1 1 2 1 1 ...
##   ..$ 47   : num [1:38] 1 1 1 1 1 1 0 1 1 1 ...
##   ..$ 48   : num [1:38] 1 1 1 1 2 1 0 1 1 1 ...
##   ..$ 49   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 50   : num [1:38] 1 1 1 1 2 1 1 1 1 1 ...
##   ..$ 51   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 52   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 53   : num [1:38] 1 1 1 1 1 1 2 1 1 1 ...
##   ..$ 54   : num [1:38] 1 1 1 1 2 1 2 2 1 1 ...
##   ..$ 55   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 56   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 57   : num [1:38] 2 2 2 1 1 2 2 2 2 1 ...
##   ..$ 58   : num [1:38] 2 2 2 1 2 2 2 2 2 1 ...
##   ..$ 59   : num [1:38] 2 2 2 1 1 2 2 2 2 1 ...
##   ..$ 60   : num [1:38] 2 2 2 1 2 2 2 2 2 1 ...
##   ..$ 61   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 62   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 63   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 64   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 65   : num [1:38] 1 1 1 1 1 1 2 1 1 1 ...
##   ..$ 66   : num [1:38] 1 1 1 1 2 1 2 1 1 1 ...
##   ..$ 67   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 68   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 69   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 70   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 71   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 72   : num [1:38] 1 1 1 1 2 1 1 1 1 1 ...
##   ..$ 73   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 74   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 75   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 76   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 77   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 78   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 79   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 80   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 81   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 82   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 83   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 84   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 85   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 86   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 87   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 88   : num [1:38] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ 89   : num [1:38] 1 2 1 1 1 1 1 1 1 1 ...
##   ..$ 90   : num [1:38] 1 2 1 1 1 1 1 1 1 1 ...
##   ..$ 91   : num [1:38] 1 2 1 1 1 1 1 1 1 1 ...
##   ..$ 92   : num [1:38] 1 2 1 1 1 1 1 1 1 1 ...
##   ..$ 93   : num [1:38] 0 1 1 1 2 1 0 1 2 2 ...
##   .. [list output truncated]
##  $ :'data.frame':    29631 obs. of  4 variables:
##   ..$ chr   : chr [1:29631] "Pf3D7_01_v3" "Pf3D7_01_v3" "Pf3D7_01_v3" "Pf3D7_01_v3" ...
##   ..$ snp_id: chr [1:29631] "Pf3D7_01_v3:93824" "Pf3D7_01_v3:94044" "Pf3D7_01_v3:94976" "Pf3D7_01_v3:95839" ...
##   ..$ pos_cM: num [1:29631] 5.47 5.49 5.54 5.59 5.61 ...
##   ..$ pos_bp: int [1:29631] 93824 94044 94976 95839 96128 96136 97100 97387 97718 98042 ...</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># reformat and filter genotypes</span>
my_genotypes &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getGenotypes.html">getGenotypes</a></span>(<span class="dt">ped.map =</span> png_pedmap,
                             <span class="dt">reference.ped.map =</span> <span class="ot">NULL</span>,
                             <span class="dt">maf =</span> <span class="fl">0.01</span>,
                             <span class="dt">isolate.max.missing =</span> <span class="fl">0.1</span>,
                             <span class="dt">snp.max.missing =</span> <span class="fl">0.1</span>,
                             <span class="dt">chromosomes =</span> <span class="ot">NULL</span>,
                             <span class="dt">input.map.distance =</span> <span class="st">"cM"</span>,
                             <span class="dt">reference.map.distance =</span> <span class="st">"cM"</span>)</code></pre></div>
<pre><code>## Begin filtering of 38 isolates and 29631 SNPs...
## 19720 SNPs remain after MAF removal...
## 18270 SNPs remain after missingness removal...</code></pre>
<pre><code>## Warning in getGenotypes(ped.map = png_pedmap, reference.ped.map = NULL, :
## isolates removed due to genotype missingness: PNG/PN0041-C</code></pre>
<pre><code>## 37 isolates remain after missingness removal...</code></pre>
<p>One isolate has been removed because more than 10% of SNPs have missing genotype calls.</p>
</div>
<div id="estimate-parameters" class="section level3">
<h3 class="hasAnchor">
<a href="#estimate-parameters" class="anchor"></a>Estimate Parameters</h3>
<p>Next we estimate the model parameters. These parameters can be useful as an initial measure of relatedness.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># estimate parameters</span>
my_parameters &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getIBDparameters.html">getIBDparameters</a></span>(<span class="dt">ped.genotypes =</span> my_genotypes, 
                                   <span class="dt">number.cores =</span> <span class="dv">1</span>)

<span class="kw">head</span>(my_parameters)</code></pre></div>
<pre><code>##   fid1     iid1 fid2     iid2        m  ibd0  ibd1 ibd2
## 1  PNG PN0003-C  PNG PN0004-C 5.608232 0.959 0.041    0
## 2  PNG PN0003-C  PNG PN0005-C 3.727380 0.849 0.151    0
## 3  PNG PN0003-C  PNG PN0006-C 4.000000 0.875 0.125    0
## 4  PNG PN0003-C  PNG PN0008-C 3.522841 0.826 0.174    0
## 5  PNG PN0003-C  PNG PN0010-C 3.878321 0.864 0.136    0
## 6  PNG PN0003-C  PNG PN0011-C 3.766112 0.853 0.147    0</code></pre>
</div>
<div id="detect-ibd-segments" class="section level3">
<h3 class="hasAnchor">
<a href="#detect-ibd-segments" class="anchor"></a>Detect IBD Segments</h3>
<p>Following parameter estimation we can detect IBD segments.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># infer IBD</span>
my_ibd &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getIBDsegments.html">getIBDsegments</a></span>(<span class="dt">ped.genotypes =</span> my_genotypes,
                         <span class="dt">parameters =</span> my_parameters, 
                         <span class="dt">number.cores =</span> <span class="dv">1</span>, 
                         <span class="dt">minimum.snps =</span> <span class="dv">20</span>, 
                         <span class="dt">minimum.length.bp =</span> <span class="dv">50000</span>,
                         <span class="dt">error =</span> <span class="fl">0.001</span>)

<span class="kw">head</span>(my_ibd)</code></pre></div>
<pre><code>##   fid1     iid1 fid2     iid2         chr           start_snp
## 1  PNG PN0003-C  PNG PN0005-C Pf3D7_07_v3  Pf3D7_07_v3:195599
## 2  PNG PN0003-C  PNG PN0008-C Pf3D7_13_v3  Pf3D7_13_v3:420993
## 3  PNG PN0003-C  PNG PN0011-C Pf3D7_06_v3 Pf3D7_06_v3:1141222
## 4  PNG PN0003-C  PNG PN0011-C Pf3D7_09_v3 Pf3D7_09_v3:1320991
## 5  PNG PN0003-C  PNG PN0022-C Pf3D7_07_v3  Pf3D7_07_v3:438905
## 6  PNG PN0003-C  PNG PN0023-C Pf3D7_09_v3   Pf3D7_09_v3:80454
##               end_snp start_position_bp end_position_bp start_position_M
## 1  Pf3D7_07_v3:483145            195599          483145       0.11411178
## 2  Pf3D7_13_v3:667638            420993          667638       0.24560586
## 3 Pf3D7_06_v3:1294070           1141222         1294070       0.66578496
## 4 Pf3D7_09_v3:1473338           1320991         1473338       0.77066157
## 5  Pf3D7_07_v3:497297            438905          497297       0.25605566
## 6  Pf3D7_09_v3:146061             80454          146061       0.04693658
##   end_position_M number_snps length_bp   length_M ibd_status
## 1     0.28186512         279    287546 0.16775334          1
## 2     0.38949770         211    246645 0.14389184          1
## 3     0.75495595         102    152848 0.08917099          1
## 4     0.85954028         128    152347 0.08887871          1
## 5     0.29012135         126     58392 0.03406569          1
## 6     0.08521148         100     65607 0.03827490          1</code></pre>
<p>We have set thresholds on the minimum number of SNPs and length of IBD segments reported in order to reduce false positive IBD calls that are most likely due to population linkage disequilibrium from extremely distant relatedness.</p>
<p>We can get a brief summary of the IBD segments as follows.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># get a summary of IBD segments</span>
<span class="kw"><a href="../reference/getIBDsummary.html">getIBDsummary</a></span>(<span class="dt">ped.genotypes =</span> my_genotypes, 
              <span class="dt">ibd.segments =</span> my_ibd) </code></pre></div>
<pre><code>## 37 / 37 isolates IBD
## 167 / 666 pairs of isolates IBD
## total IBD segments detected = 331 
## average length of segments (bp) = 492522 
## 11 IBD segments detected on chromosome Pf3D7_01_v3 
## 13 IBD segments detected on chromosome Pf3D7_02_v3 
## 10 IBD segments detected on chromosome Pf3D7_03_v3 
## 22 IBD segments detected on chromosome Pf3D7_04_v3 
## 15 IBD segments detected on chromosome Pf3D7_05_v3 
## 26 IBD segments detected on chromosome Pf3D7_06_v3 
## 61 IBD segments detected on chromosome Pf3D7_07_v3 
## 25 IBD segments detected on chromosome Pf3D7_08_v3 
## 18 IBD segments detected on chromosome Pf3D7_09_v3 
## 20 IBD segments detected on chromosome Pf3D7_10_v3 
## 22 IBD segments detected on chromosome Pf3D7_11_v3 
## 45 IBD segments detected on chromosome Pf3D7_12_v3 
## 24 IBD segments detected on chromosome Pf3D7_13_v3 
## 19 IBD segments detected on chromosome Pf3D7_14_v3</code></pre>
<p>IBD segments are found using the Viterbi algorithm, which finds the single most likely sequence of IBD states that could have generated the observed genotypic data. An alternative method to this is to calculate the posterior probability of IBD sharing, which calculates the probability of sharing 0, 1 or 2 alleles IBD at each SNP, given the genotypic data. Thus, in addition to the Viterbi algorithm, we provide a function to generate the average posterior probability of IBD sharing for each isolate pair, which is calculated as;</p>
<p><span class="math display">\[avePostPr = \frac{PostPr(IBD = 1)}{2} + PostPr(IBD = 2).\]</span> <strong>Please note, the output from the following function has number of rows equal to the number of SNPs and number of columns equal to the number of pairwise comparisons. As such, it will be very large when there are many isolates in the analysis and is likely to take some time to run. Additionally, the output may exceed R’s memory limit and, in some instances, can cause R to unexpectedly crash.</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># calculate the posterior probability of IBD sharing</span>
my_posterior &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getIBDposterior.html">getIBDposterior</a></span>(<span class="dt">ped.genotypes =</span> my_genotypes,
                         <span class="dt">parameters =</span> my_parameters, 
                         <span class="dt">number.cores =</span> <span class="dv">1</span>, 
                         <span class="dt">error =</span> <span class="fl">0.001</span>)

<span class="kw">head</span>(my_posterior[,<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>])</code></pre></div>
</div>
<div id="exploring-ibd-results" class="section level3">
<h3 class="hasAnchor">
<a href="#exploring-ibd-results" class="anchor"></a>Exploring IBD Results</h3>
<p>isoRelate provides a number of graphical functions to help with exploring the vast number of IBD segments detected. We begin by plotting the IBD segments as colored blocks across the genome to get an idea of their size and distribution.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># plot IBD segments</span>
<span class="kw"><a href="../reference/plotIBDsegments.html">plotIBDsegments</a></span>(<span class="dt">ped.genotypes =</span> my_genotypes, 
                <span class="dt">ibd.segments =</span> my_ibd, 
                <span class="dt">interval =</span> <span class="ot">NULL</span>,
                <span class="dt">annotation.genes =</span> <span class="ot">NULL</span>,
                <span class="dt">annotation.genes.color =</span> <span class="ot">NULL</span>,
                <span class="dt">highlight.genes =</span> <span class="ot">NULL</span>, 
                <span class="dt">highlight.genes.labels =</span> <span class="ot">FALSE</span>,
                <span class="dt">highlight.genes.color =</span> <span class="ot">NULL</span>,
                <span class="dt">highlight.genes.alpha =</span> <span class="fl">0.1</span>,
                <span class="dt">segment.height =</span> <span class="fl">0.6</span>,
                <span class="dt">number.per.page =</span> <span class="ot">NULL</span>, 
                <span class="dt">fid.label =</span> <span class="ot">FALSE</span>, 
                <span class="dt">iid.label =</span> <span class="ot">FALSE</span>, 
                <span class="dt">ylabel.size =</span> <span class="dv">9</span>, 
                <span class="dt">add.rug =</span> <span class="ot">FALSE</span>,
                <span class="dt">plot.title =</span> <span class="st">"Distribution of IBD segments in PNG"</span>, 
                <span class="dt">add.legend =</span> <span class="ot">FALSE</span>, 
                <span class="dt">segment.color =</span> <span class="ot">NULL</span>)</code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-9-1.png" width="672"></p>
<p>If there are many IBD pairs then it may be worthwhile specifying the maximum number of pairs to plot in a figure (<code>number.per.page</code>). This will split all IBD pairs into subgroups of IBD pairs of size <code>number.per.page</code>, then plot the IBD segments for each subgroup in a new figure. It is <strong>highly</strong> recommended to save the figures directly to a document (<code>my_ibd_segments.pdf</code>) as opposed to the R plot window if there are many IBD pairs.</p>
<p>We have reframed from including isolate labels (y-axis) in this plot. Isolate pairs who are highly related will have many IBD segments spanning a large amount of the genome. Identical isolates will have IBD blocks spanning the entire genome.</p>
<p>We detected many IBD segments on chromosome 7 (Pf3D7_07_v3), suggesting a loci on chromosome 7 may be under selection. We can take a closer look at this region by specifying an interval to plot the segments over as well as including annotation genes or genes that may be of interest over this region. We alter some of the parameters for fun.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># plot IBD segments over an interval: chromosome 7: 350000 - 550000</span>
<span class="kw"><a href="../reference/plotIBDsegments.html">plotIBDsegments</a></span>(<span class="dt">ped.genotypes =</span> my_genotypes,
                <span class="dt">ibd.segments =</span> my_ibd,
                <span class="dt">interval =</span> <span class="kw">c</span>(<span class="st">"Pf3D7_07_v3"</span>,<span class="dv">350000</span>,<span class="dv">550000</span>),
                <span class="dt">annotation.genes =</span> annotation_genes,
                <span class="dt">annotation.genes.color =</span> <span class="ot">NULL</span>,
                <span class="dt">highlight.genes =</span> highlight_genes,
                <span class="dt">highlight.genes.labels =</span> <span class="ot">FALSE</span>,
                <span class="dt">highlight.genes.color =</span> <span class="ot">NULL</span>,
                <span class="dt">highlight.genes.alpha =</span> <span class="fl">0.1</span>,
                <span class="dt">segment.height =</span> <span class="fl">0.8</span>,
                <span class="dt">number.per.page =</span> <span class="ot">NULL</span>,
                <span class="dt">fid.label =</span> <span class="ot">FALSE</span>,
                <span class="dt">iid.label =</span> <span class="ot">FALSE</span>,
                <span class="dt">ylabel.size =</span> <span class="dv">9</span>,
                <span class="dt">add.rug =</span> <span class="ot">TRUE</span>,
                <span class="dt">plot.title =</span> <span class="st">"Distribution of IBD segments in PNG"</span>,
                <span class="dt">add.legend =</span> <span class="ot">FALSE</span>,
                <span class="dt">segment.color =</span> <span class="kw">c</span>(<span class="st">"purple"</span>,<span class="st">"green"</span>))</code></pre></div>
<pre><code>## Warning: Removed 1 rows containing missing values (geom_rect).</code></pre>
<p><img src="introduction_files/figure-html/unnamed-chunk-10-1.png" width="672"></p>
<p>The gene CRT (<em>Plasmodium falciparum</em> chloroquine resistance transporter) overlaps many of the segments. Mutations in this gene have been associated with antimalarial drug resistance, resulting in positive selection of this gene and hence many IBD segments. Ticks on the x-axis inside the plotting area indicate SNP positions (<code>add.rug=TRUE</code>) while yellow and red colored segments represent annotation genes (positive strand and negative strand respectively).</p>
<p>As an alternative to plotting the IBD segments as above, which may become unfeasible when there are many IBD pairs, we can calculate the proportion of pairs who are IBD at each SNP and plot this distribution across the genome as a quick way of identifying genomic loci with many IBD pairs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># generate a binary IBD matrix</span>
my_matrix &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getIBDmatrix.html">getIBDmatrix</a></span>(<span class="dt">ped.genotypes =</span> my_genotypes, 
                          <span class="dt">ibd.segments =</span> my_ibd)

<span class="co"># calculate the proportion of pairs IBD at each SNP</span>
my_proportion &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getIBDproportion.html">getIBDproportion</a></span>(<span class="dt">ped.genotypes =</span> my_genotypes, 
                                  <span class="dt">ibd.matrix =</span> my_matrix, 
                                  <span class="dt">groups =</span> <span class="ot">NULL</span>)

<span class="co"># plot the proportion of pairs IBD</span>
<span class="kw"><a href="../reference/plotIBDproportions.html">plotIBDproportions</a></span>(<span class="dt">ibd.proportions =</span> my_proportion, 
                   <span class="dt">interval =</span> <span class="ot">NULL</span>, 
                   <span class="dt">annotation.genes =</span> <span class="ot">NULL</span>,
                   <span class="dt">annotation.genes.color =</span> <span class="ot">NULL</span>,
                   <span class="dt">highlight.genes =</span> <span class="ot">NULL</span>,
                   <span class="dt">highlight.genes.labels =</span> <span class="ot">TRUE</span>,
                   <span class="dt">highlight.genes.color =</span> <span class="ot">NULL</span>,
                   <span class="dt">highlight.genes.alpha =</span> <span class="fl">0.1</span>,
                   <span class="dt">add.rug =</span> <span class="ot">FALSE</span>, 
                   <span class="dt">plot.title =</span> <span class="st">"Proportion of pairs IBD in PNG"</span>, 
                   <span class="dt">add.legend =</span> <span class="ot">FALSE</span>,
                   <span class="dt">line.color =</span> <span class="ot">NULL</span>, 
                   <span class="dt">facet.label =</span> <span class="ot">TRUE</span>, 
                   <span class="dt">facet.scales =</span> <span class="st">"fixed"</span>, 
                   <span class="dt">subpop.facet =</span> <span class="ot">FALSE</span>)</code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-11-1.png" width="672"></p>
<p>The region on chromosome 7 has a high proportion of pairs that are IBD, potentially reflective of positive selection. This figure can be further explored if additional information is available. For example, we could stratify the results by collection sites to see if different sites have different loci potentially under selection. All of the PNG isolates were collected from Madang and no other location, therefore we create an example grouping of all the isolates for stratification purposes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># creating a stratification dataset </span>
my_groups &lt;-<span class="st"> </span>my_genotypes[[<span class="dv">1</span>]][,<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]
my_groups[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>,<span class="st">"pid"</span>] &lt;-<span class="st"> "a"</span>
my_groups[<span class="dv">11</span><span class="op">:</span><span class="dv">25</span>,<span class="st">"pid"</span>] &lt;-<span class="st"> "b"</span>
my_groups[<span class="dv">26</span><span class="op">:</span><span class="dv">38</span>,<span class="st">"pid"</span>] &lt;-<span class="st"> "c"</span>

my_proportion &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getIBDproportion.html">getIBDproportion</a></span>(<span class="dt">ped.genotypes =</span> my_genotypes,
                                  <span class="dt">ibd.matrix =</span> my_matrix,
                                  <span class="dt">groups =</span> my_groups)

<span class="co"># plot the proportion of pairs IBD</span>
<span class="kw"><a href="../reference/plotIBDproportions.html">plotIBDproportions</a></span>(<span class="dt">ibd.proportions =</span> my_proportion,
                   <span class="dt">interval =</span> <span class="ot">NULL</span>,
                   <span class="dt">annotation.genes =</span> <span class="ot">NULL</span>,
                   <span class="dt">annotation.genes.color =</span> <span class="ot">NULL</span>,
                   <span class="dt">highlight.genes =</span> <span class="ot">NULL</span>,
                   <span class="dt">highlight.genes.labels =</span> <span class="ot">FALSE</span>,
                   <span class="dt">highlight.genes.color =</span> <span class="ot">NULL</span>,
                   <span class="dt">highlight.genes.alpha =</span> <span class="fl">0.1</span>,
                   <span class="dt">line.color =</span> <span class="ot">NULL</span>,
                   <span class="dt">add.rug =</span> <span class="ot">FALSE</span>,
                   <span class="dt">plot.title =</span> <span class="st">"Proportion of pairs IBD in PNG - with stratification"</span>,
                   <span class="dt">add.legend =</span> <span class="ot">FALSE</span>,
                   <span class="dt">facet.label =</span> <span class="ot">TRUE</span>,
                   <span class="dt">facet.scales =</span> <span class="st">"fixed"</span>,
                   <span class="dt">subpop.facet =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-12-1.png" width="672"></p>
<p>To identify genomic loci with significant amounts of excess IBD we can apply a transformation to the binary IBD matrix to account for variations in isolate relatedness as well as SNP allele frequencies, then calculate a summary statistic that can be used to assess significance.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># calculate the significance of IBD sharing</span>
my_iR &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getIBDiR.html">getIBDiR</a></span>(<span class="dt">ped.genotypes =</span> my_genotypes, 
                  <span class="dt">ibd.matrix =</span> my_matrix, 
                  <span class="dt">groups =</span> <span class="ot">NULL</span>)

<span class="co"># plot the iR statistics</span>
<span class="kw"><a href="../reference/plotIBDiR.html">plotIBDiR</a></span>(<span class="dt">ibd.iR =</span> my_iR, 
          <span class="dt">interval =</span> <span class="ot">NULL</span>, 
          <span class="dt">annotation.genes =</span> <span class="ot">NULL</span>,
          <span class="dt">annotation.genes.color =</span> <span class="ot">NULL</span>,
          <span class="dt">highlight.genes =</span> <span class="ot">NULL</span>,
          <span class="dt">highlight.genes.labels =</span> <span class="ot">FALSE</span>,
          <span class="dt">highlight.genes.color =</span> <span class="ot">NULL</span>,
          <span class="dt">highlight.genes.alpha =</span> <span class="fl">0.1</span>,
          <span class="dt">point.size =</span> <span class="dv">1</span>,
          <span class="dt">point.color =</span> <span class="ot">NULL</span>,
          <span class="dt">add.rug =</span> <span class="ot">FALSE</span>, 
          <span class="dt">plot.title =</span> <span class="st">"Significance of IBD sharing"</span>, 
          <span class="dt">add.legend =</span> <span class="ot">FALSE</span>,
          <span class="dt">facet.label =</span> <span class="ot">TRUE</span>, 
          <span class="dt">facet.scales =</span> <span class="st">"fixed"</span>)</code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-13-1.png" width="672"></p>
<p>A 5% significance threshold would identify regions that have SNPs with -log10 (P-values) &gt; -log10(0.05) as having significant IBD sharing, indicative of positive selection.</p>
<p>We note that the function <code>getIBDiR</code> can return <code>NA</code> values which may occur when there are no IBD pairs or when all pairs are IBD, or when only several isolates are analyzed, among other reasons.</p>
<p>We can investigate the specific isolate pairs that are contributing to the excess IBD sharing on chromosome 7 by creating an IBD network. Such a network will identify clusters of isolates sharing a common haplotype over this region. We include our example grouping of isolates when plotting the network.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># generate the isolates who are IBD over the Plasmodium falciparum CRT gene</span>
my_i_clusters &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getIBDiclusters.html">getIBDiclusters</a></span>(<span class="dt">ped.genotypes =</span> my_genotypes, 
                                 <span class="dt">ibd.segments =</span> my_ibd, 
                                 <span class="dt">interval =</span> <span class="kw">c</span>(<span class="st">"Pf3D7_07_v3"</span>, <span class="dv">403222</span>, <span class="dv">406317</span>), 
                                 <span class="dt">prop=</span><span class="dv">0</span>, 
                                 <span class="dt">hi.clust =</span> <span class="ot">FALSE</span>)</code></pre></div>
<pre><code>## Number of clusters = 5
## Maximum cluster length = 9
## Minimum cluster length = 2</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(my_i_clusters)</code></pre></div>
<pre><code>## List of 3
##  $ clusters :List of 5
##   ..$ : chr [1:9] "PNG/PN0014-C" "PNG/PN0023-C" "PNG/PN0031-C" "PNG/PN0040-C" ...
##   ..$ : chr [1:5] "PNG/PN0057-C" "PNG/PN0078-C" "PNG/PN0079-C" "PNG/PN0086-C" ...
##   ..$ : chr [1:3] "PNG/PN0003-C" "PNG/PN0005-C" "PNG/PN0083-C"
##   ..$ : chr [1:2] "PNG/PN0022-C" "PNG/PN0042-C"
##   ..$ : chr [1:2] "PNG/PN0037-C" "PNG/PN0058-Cx"
##  $ i.network:List of 10
##   ..$ :List of 1
##   .. ..$ PNG/PN0003-C:Class 'igraph.vs'  atomic [1:1] 2
##   .. .. .. ..- attr(*, "env")=&lt;weakref&gt; 
##   .. .. .. ..- attr(*, "graph")= chr "0bb3b7d6-e2fb-4773-a7fd-134f8573569a"
##   ..$ :List of 1
##   .. ..$ PNG/PN0005-C:Class 'igraph.vs'  atomic [1:2] 1 17
##   .. .. .. ..- attr(*, "env")=&lt;weakref&gt; 
##   .. .. .. ..- attr(*, "graph")= chr "0bb3b7d6-e2fb-4773-a7fd-134f8573569a"
##   ..$ :List of 1
##   .. ..$ PNG/PN0014-C:Class 'igraph.vs'  atomic [1:8] 5 6 8 9 10 11 13 18
##   .. .. .. ..- attr(*, "env")=&lt;weakref&gt; 
##   .. .. .. ..- attr(*, "graph")= chr "0bb3b7d6-e2fb-4773-a7fd-134f8573569a"
##   ..$ :List of 1
##   .. ..$ PNG/PN0022-C:Class 'igraph.vs'  atomic [1:1] 19
##   .. .. .. ..- attr(*, "env")=&lt;weakref&gt; 
##   .. .. .. ..- attr(*, "graph")= chr "0bb3b7d6-e2fb-4773-a7fd-134f8573569a"
##   ..$ :List of 1
##   .. ..$ PNG/PN0023-C:Class 'igraph.vs'  atomic [1:7] 3 6 8 9 10 11 13
##   .. .. .. ..- attr(*, "env")=&lt;weakref&gt; 
##   .. .. .. ..- attr(*, "graph")= chr "0bb3b7d6-e2fb-4773-a7fd-134f8573569a"
##   ..$ :List of 1
##   .. ..$ PNG/PN0031-C:Class 'igraph.vs'  atomic [1:7] 3 5 8 9 10 11 13
##   .. .. .. ..- attr(*, "env")=&lt;weakref&gt; 
##   .. .. .. ..- attr(*, "graph")= chr "0bb3b7d6-e2fb-4773-a7fd-134f8573569a"
##   ..$ :List of 1
##   .. ..$ PNG/PN0037-C:Class 'igraph.vs'  atomic [1:1] 20
##   .. .. .. ..- attr(*, "env")=&lt;weakref&gt; 
##   .. .. .. ..- attr(*, "graph")= chr "0bb3b7d6-e2fb-4773-a7fd-134f8573569a"
##   ..$ :List of 1
##   .. ..$ PNG/PN0040-C:Class 'igraph.vs'  atomic [1:8] 3 5 6 9 10 11 13 18
##   .. .. .. ..- attr(*, "env")=&lt;weakref&gt; 
##   .. .. .. ..- attr(*, "graph")= chr "0bb3b7d6-e2fb-4773-a7fd-134f8573569a"
##   ..$ :List of 1
##   .. ..$ PNG/PN0055-C:Class 'igraph.vs'  atomic [1:7] 3 5 6 8 10 11 13
##   .. .. .. ..- attr(*, "env")=&lt;weakref&gt; 
##   .. .. .. ..- attr(*, "graph")= chr "0bb3b7d6-e2fb-4773-a7fd-134f8573569a"
##   ..$ :List of 1
##   .. ..$ PNG/PN0055-Cx:Class 'igraph.vs'  atomic [1:7] 3 5 6 8 9 11 13
##   .. .. .. ..- attr(*, "env")=&lt;weakref&gt; 
##   .. .. .. ..- attr(*, "graph")= chr "0bb3b7d6-e2fb-4773-a7fd-134f8573569a"
##   ..- attr(*, "class")= chr "igraph"
##  $ hi.clust : logi FALSE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># plot the network of clusters</span>
<span class="kw"><a href="../reference/plotIBDclusters.html">plotIBDclusters</a></span>(<span class="dt">ped.genotypes =</span> my_genotypes, 
                <span class="dt">clusters =</span> my_i_clusters, 
                <span class="dt">groups =</span> my_groups, 
                <span class="dt">vertex.color =</span> <span class="ot">NULL</span>, 
                <span class="dt">vertex.frame.color =</span> <span class="st">"white"</span>,
                <span class="dt">vertex.size =</span> <span class="dv">4</span>, 
                <span class="dt">vertex.name =</span> <span class="ot">FALSE</span>, 
                <span class="dt">edge.color =</span> <span class="st">"gray60"</span>, 
                <span class="dt">edge.width =</span> <span class="fl">0.8</span>, 
                <span class="dt">mark.border =</span> <span class="st">"white"</span>,
                <span class="dt">mark.col =</span> <span class="st">"gray94"</span>, 
                <span class="dt">add.legend =</span> <span class="ot">TRUE</span>, 
                <span class="dt">legend.x =</span> <span class="op">-</span><span class="fl">1.5</span>, 
                <span class="dt">legend.y =</span> <span class="op">-</span><span class="fl">0.25</span>, 
                <span class="dt">layout =</span> <span class="ot">NULL</span>, 
                <span class="dt">return.layout =</span> <span class="ot">FALSE</span>)</code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-14-1.png" width="672"></p>
<p>Each time the function <code><a href="../reference/plotIBDclusters.html">plotIBDclusters()</a></code> is run the vertex layout will differ. This is an artifact of the R package used to generate the network, <code>igraph</code>. If the user wishes to keep the network layout the same in multiple subsequent figures then set <code>return.layout = TRUE</code> and use the returned layout in subsequent <code><a href="../reference/plotIBDclusters.html">plotIBDclusters()</a></code> runs with <code>layout = my_layout</code>.</p>
<p>We note that the <code>i.network</code> is output as the second object in the list <code>my_i_clusters</code>. If the user wishes to perform quantitative analyses using this network, there are many functions in the <code>i.graph</code> package to do so. Please see the <code>i.graph</code> documentation for more details on this.</p>
<p>We can also explore networks of isolates sharing a large proportion of their genome IBD. Such isolates may represent duplicate samples or individuals who have identical infections.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># generate the isolates who share at least than 90% of their genome IBD</span>
my_p_clusters &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getIBDpclusters.html">getIBDpclusters</a></span>(<span class="dt">ped.genotypes =</span> my_genotypes, 
                                 <span class="dt">ibd.segments =</span> my_ibd, 
                                 <span class="dt">prop=</span><span class="fl">0.9</span>, 
                                 <span class="dt">hi.clust =</span> <span class="ot">FALSE</span>)</code></pre></div>
<pre><code>## Number of clusters = 3
## Maximum cluster length = 3
## Minimum cluster length = 2</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># plot the network of clusters</span>
<span class="kw"><a href="../reference/plotIBDclusters.html">plotIBDclusters</a></span>(<span class="dt">ped.genotypes =</span> my_genotypes, 
                <span class="dt">clusters =</span> my_p_clusters, 
                <span class="dt">groups =</span> my_groups, 
                <span class="dt">vertex.color =</span> <span class="ot">NULL</span>, 
                <span class="dt">vertex.frame.color =</span> <span class="st">"white"</span>,
                <span class="dt">vertex.size =</span> <span class="dv">4</span>, 
                <span class="dt">vertex.name =</span> <span class="ot">FALSE</span>, 
                <span class="dt">edge.color =</span> <span class="st">"gray60"</span>, 
                <span class="dt">edge.width =</span> <span class="fl">0.8</span>, 
                <span class="dt">mark.border =</span> <span class="st">"white"</span>,
                <span class="dt">mark.col =</span> <span class="st">"gray94"</span>, 
                <span class="dt">add.legend =</span> <span class="ot">TRUE</span>, 
                <span class="dt">legend.x =</span> <span class="op">-</span><span class="fl">1.5</span>, 
                <span class="dt">legend.y =</span> <span class="op">-</span><span class="fl">0.25</span>, 
                <span class="dt">layout =</span> <span class="ot">NULL</span>, 
                <span class="dt">return.layout =</span> <span class="ot">FALSE</span>)</code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-15-1.png" width="672"></p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#data-format">Data Format</a></li>
      <li><a href="#key-functions">Key Functions</a></li>
      <li><a href="#example">Example</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Lyndal Henden.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
