<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Interval Cluster Networks — getIBDiclusters • isoRelate</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  
  
<meta property="og:title" content="Interval Cluster Networks — getIBDiclusters" />

<meta property="og:description" content="getIBDiclusters() produces a network of clusters of isolates that have been inferred IBD over a specified interval.
Isolates that are not IBD over the interval are not included in the network or output. The networks are created
using the R package igraph." />
<meta name="twitter:card" content="summary" />
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">isoRelate</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/introduction.html">isoRelate Introduction</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Interval Cluster Networks</h1>
    </div>

    
    <p><code>getIBDiclusters()</code> produces a network of clusters of isolates that have been inferred IBD over a specified interval.
Isolates that are not IBD over the interval are not included in the network or output. The networks are created
using the R package <code>igraph</code>.</p>
    

    <pre class="usage"><span class='fu'>getIBDiclusters</span>(<span class='no'>ped.genotypes</span>, <span class='no'>ibd.segments</span>, <span class='kw'>interval</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>prop</span> <span class='kw'>=</span> <span class='fl'>0</span>,
  <span class='kw'>hi.clust</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>ped.genotypes</th>
      <td><p>A list containing 2 objects. See the <code>Value</code> description in <code><a href='getGenotypes.html'>getGenotypes</a></code> for more details on this input.</p></td>
    </tr>
    <tr>
      <th>ibd.segments</th>
      <td><p>A data frame containing the IBD segments detected by isoRelate.
See the <code>Value</code> description in <code><a href='getIBDsegments.html'>getIBDsegments</a></code> for more details on this input.</p></td>
    </tr>
    <tr>
      <th>interval</th>
      <td><p>A vector of length 3 containing the region to identify clusters over. This vector should contain the
chromosome ID, the start of the interval in base-pairs and the end of the interval in base-pairs; in this order respectively.</p></td>
    </tr>
    <tr>
      <th>prop</th>
      <td><p>Numeric value between 0 and 1 (inclusive).
The minimum proportion of an interval (in base-pairs) shared IBD between a pair of isolates in order for the pair to be included in the network.
For example, if <code>prop=1</code> then two isolates will be included if they are IBD over the entire interval,
whereas <code>prop=0.5</code> will include isolates that are IBD over at least 50% of the interval.
The default is <code>prop=0</code>, which includes isolates if they have an IBD segment that overlaps the interval in any way.</p></td>
    </tr>
    <tr>
      <th>hi.clust</th>
      <td><p>Logical. Whether to perform hierarchical clustering using the <code>fastgreedy.community</code> approach in the <code>igraph</code> package.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list of three objects named <code>clusters</code>, <code>i.network</code> and <code>hi.clust</code>:</p><ol>
<li><p>A list where each object contains the names of isolates that form a disjoint cluster in the network.
If hierarchical clustering has been performed then the clusters may not be disjoint.</p></li>
<li><p>An <code>igraph</code> network used in the construction of network plots. See <a href=' http://igraph.org/r/'> http://igraph.org/r/</a> doe more details.</p></li>
<li><p>Logical. Whether or not hierarchical clustering has been performed.</p></li>
</ol>

    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <p><code><a href='getGenotypes.html'>getGenotypes</a></code>, <code><a href='getIBDsegments.html'>getIBDsegments</a></code> and <code><a href='getIBDpclusters.html'>getIBDpclusters</a></code>.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># generate the isolates who are IBD over the Plasmodium falciparum CRT gene</span>
<span class='no'>my_i_clusters</span> <span class='kw'>&lt;-</span> <span class='fu'>getIBDiclusters</span>(<span class='kw'>ped.genotypes</span> <span class='kw'>=</span> <span class='no'>png_genotypes</span>,
                                 <span class='kw'>ibd.segments</span> <span class='kw'>=</span> <span class='no'>png_ibd</span>,
                                 <span class='kw'>interval</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"Pf3D7_07_v3"</span>, <span class='fl'>403222</span>, <span class='fl'>406317</span>),
                                 <span class='kw'>prop</span><span class='kw'>=</span><span class='fl'>0</span>,
                                 <span class='kw'>hi.clust</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</div><div class='output co'>#&gt; Number of clusters = 5
#&gt; Maximum cluster length = 9
#&gt; Minimum cluster length = 2</div><div class='input'>
<span class='fu'>str</span>(<span class='no'>my_i_clusters</span>)</div><div class='output co'>#&gt; List of 3
#&gt;  $ clusters :List of 5
#&gt;   ..$ : chr [1:9] "PNG/PN0014-C" "PNG/PN0023-C" "PNG/PN0031-C" "PNG/PN0040-C" ...
#&gt;   ..$ : chr [1:5] "PNG/PN0057-C" "PNG/PN0078-C" "PNG/PN0079-C" "PNG/PN0086-C" ...
#&gt;   ..$ : chr [1:3] "PNG/PN0003-C" "PNG/PN0005-C" "PNG/PN0083-C"
#&gt;   ..$ : chr [1:2] "PNG/PN0022-C" "PNG/PN0042-C"
#&gt;   ..$ : chr [1:2] "PNG/PN0037-C" "PNG/PN0058-Cx"
#&gt;  $ i.network:List of 10
#&gt;   ..$ :List of 1
#&gt;   .. ..$ PNG/PN0003-C:Class 'igraph.vs'  atomic [1:1] 2
#&gt;   .. .. .. ..- attr(*, "env")=&lt;weakref&gt; 
#&gt;   .. .. .. ..- attr(*, "graph")= chr "39a4976c-6a81-4e3a-86a3-5fef0eb92dbe"
#&gt;   ..$ :List of 1
#&gt;   .. ..$ PNG/PN0005-C:Class 'igraph.vs'  atomic [1:2] 1 17
#&gt;   .. .. .. ..- attr(*, "env")=&lt;weakref&gt; 
#&gt;   .. .. .. ..- attr(*, "graph")= chr "39a4976c-6a81-4e3a-86a3-5fef0eb92dbe"
#&gt;   ..$ :List of 1
#&gt;   .. ..$ PNG/PN0014-C:Class 'igraph.vs'  atomic [1:8] 5 6 8 9 10 11 13 18
#&gt;   .. .. .. ..- attr(*, "env")=&lt;weakref&gt; 
#&gt;   .. .. .. ..- attr(*, "graph")= chr "39a4976c-6a81-4e3a-86a3-5fef0eb92dbe"
#&gt;   ..$ :List of 1
#&gt;   .. ..$ PNG/PN0022-C:Class 'igraph.vs'  atomic [1:1] 19
#&gt;   .. .. .. ..- attr(*, "env")=&lt;weakref&gt; 
#&gt;   .. .. .. ..- attr(*, "graph")= chr "39a4976c-6a81-4e3a-86a3-5fef0eb92dbe"
#&gt;   ..$ :List of 1
#&gt;   .. ..$ PNG/PN0023-C:Class 'igraph.vs'  atomic [1:7] 3 6 8 9 10 11 13
#&gt;   .. .. .. ..- attr(*, "env")=&lt;weakref&gt; 
#&gt;   .. .. .. ..- attr(*, "graph")= chr "39a4976c-6a81-4e3a-86a3-5fef0eb92dbe"
#&gt;   ..$ :List of 1
#&gt;   .. ..$ PNG/PN0031-C:Class 'igraph.vs'  atomic [1:7] 3 5 8 9 10 11 13
#&gt;   .. .. .. ..- attr(*, "env")=&lt;weakref&gt; 
#&gt;   .. .. .. ..- attr(*, "graph")= chr "39a4976c-6a81-4e3a-86a3-5fef0eb92dbe"
#&gt;   ..$ :List of 1
#&gt;   .. ..$ PNG/PN0037-C:Class 'igraph.vs'  atomic [1:1] 20
#&gt;   .. .. .. ..- attr(*, "env")=&lt;weakref&gt; 
#&gt;   .. .. .. ..- attr(*, "graph")= chr "39a4976c-6a81-4e3a-86a3-5fef0eb92dbe"
#&gt;   ..$ :List of 1
#&gt;   .. ..$ PNG/PN0040-C:Class 'igraph.vs'  atomic [1:8] 3 5 6 9 10 11 13 18
#&gt;   .. .. .. ..- attr(*, "env")=&lt;weakref&gt; 
#&gt;   .. .. .. ..- attr(*, "graph")= chr "39a4976c-6a81-4e3a-86a3-5fef0eb92dbe"
#&gt;   ..$ :List of 1
#&gt;   .. ..$ PNG/PN0055-C:Class 'igraph.vs'  atomic [1:7] 3 5 6 8 10 11 13
#&gt;   .. .. .. ..- attr(*, "env")=&lt;weakref&gt; 
#&gt;   .. .. .. ..- attr(*, "graph")= chr "39a4976c-6a81-4e3a-86a3-5fef0eb92dbe"
#&gt;   ..$ :List of 1
#&gt;   .. ..$ PNG/PN0055-Cx:Class 'igraph.vs'  atomic [1:7] 3 5 6 8 9 11 13
#&gt;   .. .. .. ..- attr(*, "env")=&lt;weakref&gt; 
#&gt;   .. .. .. ..- attr(*, "graph")= chr "39a4976c-6a81-4e3a-86a3-5fef0eb92dbe"
#&gt;   ..- attr(*, "class")= chr "igraph"
#&gt;  $ hi.clust : logi FALSE</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Lyndal Henden.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>

