<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Selection Significance Statistic — getIBDiR • isoRelate</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  
  
<meta property="og:title" content="Selection Significance Statistic — getIBDiR" />

<meta property="og:description" content="getIBDiR() calculates a summary statistic for each SNP that can be used to assess the significance of excess IBD sharing at genomic loci,
thus identifying regions under positive selection.
First relatedness between isolates and SNP allele frequencies are accounted for, then normalization procedures are applied where we assume
our transformed summary statistic follows a chi-squared distribution with 1 degree of freedom.
This allows the calculation of -log10 (P-values) which we denote as the iR statistic.
SNPs with iR values greater than some threshold (i.e. -log10 (P-values) &gt; -log10 (0.05)) provide evidence of positive selection.
getIBDiR can return NA iR statistics for a number of reasons, including trying to generate iR statistics when there are no IBD pairs
or when all pairs are IBD, or when only several isolates are analyzed." />
<meta name="twitter:card" content="summary" />
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">isoRelate</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/introduction.html">isoRelate Introduction</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Selection Significance Statistic</h1>
    </div>

    
    <p><code>getIBDiR()</code> calculates a summary statistic for each SNP that can be used to assess the significance of excess IBD sharing at genomic loci,
thus identifying regions under positive selection.
First relatedness between isolates and SNP allele frequencies are accounted for, then normalization procedures are applied where we assume
our transformed summary statistic follows a chi-squared distribution with 1 degree of freedom.
This allows the calculation of -log10 (P-values) which we denote as the iR statistic.
SNPs with iR values greater than some threshold (i.e. -log10 (P-values) &gt; -log10 (0.05)) provide evidence of positive selection.
<code>getIBDiR</code> can return <code>NA</code> iR statistics for a number of reasons, including trying to generate iR statistics when there are no IBD pairs
or when all pairs are IBD, or when only several isolates are analyzed.</p>
    

    <pre class="usage"><span class='fu'>getIBDiR</span>(<span class='no'>ped.genotypes</span>, <span class='no'>ibd.matrix</span>, <span class='kw'>groups</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>ped.genotypes</th>
      <td><p>A list containing 2 objects. See the <code>Value</code> description in <code><a href='getGenotypes.html'>getGenotypes</a></code> for more details on this input.</p></td>
    </tr>
    <tr>
      <th>ibd.matrix</th>
      <td><p>A data frame containing the binary IBD information for each SNP and each pair.
See the returned <code>Value</code> in <code><a href='getIBDmatrix.html'>getIBDmatrix</a></code> for more details.</p></td>
    </tr>
    <tr>
      <th>groups</th>
      <td><p>A data frame with 3 columns of information:</p><ol>
<li><p>Family ID</p></li>
<li><p>Isolate ID</p></li>
<li><p>Group ID</p></li>
</ol><p>where IBD proportions are calculated for</p><ol>
<li><p>all pairs of isolates within the same group</p></li>
<li><p>all pairwise-group comparisons where isolates belong to different groups</p></li>
</ol><p>Group ID, for example, can be the geographic regions where the isolates were collected.
The default is <code>groups=NULL</code> and IBD proportions will be calculated over all pairs.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A data frame the following 7 columns:</p><ol>
<li><p>Chromosome (type <code>"character"</code>, <code>"numeric"</code> or <code>"integer"</code>)</p></li>
<li><p>SNP identifiers (type <code>"character"</code>)</p></li>
<li><p>Genetic map distance (centi morgans, cM) (type <code>"numeric"</code>)</p></li>
<li><p>Base-pair position (type <code>"integer"</code>)</p></li>
<li><p>Population (type <code>"character"</code> or <code>"numeric"</code>)</p></li>
<li><p>Subpopulation (type <code>"character"</code> or <code>"numeric"</code>)</p></li>
<li><p>iR statistic (type <code>"numeric"</code>)</p></li>
<li><p>-log10 p vlaue (type <code>"numeric"</code>)</p></li>
</ol><p>where each row describes a unique SNP.
The column <code>Population</code> is filled with 1s by default, while <code>Subpopulation</code> contains the group IDs from <code>groups</code>,
where the proportion of pairs IBD has been calculated for all pairs of isolates belonging to the same group as well as all pairs of
isolates where each isolate belongs to a different group.
If <code>groups=NULL</code> then <code>Subpopulation</code> will be filled with 0s also.
The population columns have been included for plotting purposes.
The data frame is headed <code>chr, snp_id, pos_M, pos_bp, pop, subpop, iR</code> and <code>log10_pvalue</code> respectively.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <p><code><a href='getGenotypes.html'>getGenotypes</a></code>, <code><a href='getIBDmatrix.html'>getIBDmatrix</a></code> and <code><a href='getIBDproportion.html'>getIBDproportion</a></code>.</p>
    

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#see-also">See also</a></li>
          </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Lyndal Henden.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>

