<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ isoRelate</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  
  
<meta property="og:title" content="Function reference" />
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">isoRelate</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/introduction.html">isoRelate Introduction</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9">
    <div class="page-header">
      <h1>
        Reference
        <small>version&nbsp;0.1.0</small>
      </h1>
    </div>

    <div class="contents">
      <table class="ref-index">

      <colgroup>
        <col class="alias" />
        <col class="title" />
      </colgroup>

      <tbody>
        <tr>
          <th colspan="2">
            <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
            <p class="section-desc"></p>
          </th>
        </tr>
        <tr>
          <!--  -->
          <td>
            <p><code><a href="AmatrixDD.html">AmatrixDD</a></code> </p>
          </td>
          <td><p>The matrix A in the equation Ax=b for 2 diploid chromosomes</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="AmatrixHD.html">AmatrixHD</a></code> </p>
          </td>
          <td><p>The matrix A in the equation Ax=b for 1 haploid and 1 diploid chromosome</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="AmatrixHH.html">AmatrixHH</a></code> </p>
          </td>
          <td><p>The matrix A in the equation Ax=b for 2 haploid chromosomes</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="IBDLabel.html">IBDLabel</a></code> </p>
          </td>
          <td><p>Internal Function</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="IBDMatrix.html">IBDMatrix</a></code> </p>
          </td>
          <td><p>Binary Matrix of IBD</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="IBDTable.html">IBDTable</a></code> </p>
          </td>
          <td><p>Internal Function</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="IBDparameters.html">IBDparameters</a></code> </p>
          </td>
          <td><p>IBDparameters() calculates IBD probabilities then estimates the number of meiosis for an isolate pair.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="annotation_genes.html">annotation_genes</a></code> </p>
          </td>
          <td><p>Plasmodium Falciparum Gene Annotation Dataset</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="areColors.html">areColors</a></code> </p>
          </td>
          <td><p>Internal function</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="bVectorDD.html">bVectorDD</a></code> </p>
          </td>
          <td><p>The vector b in the equation Ax=b for 2 diploid chromosomes</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="bVectorHD.html">bVectorHD</a></code> </p>
          </td>
          <td><p>The vector b in the equation Ax=b for 1 haploid chromosome and 1 diploid chromosome</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="bVectorHH.html">bVectorHH</a></code> </p>
          </td>
          <td><p>The vector b in the equation Ax=b for 2 haploid chromosomes</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="calculateAlpha.html">calculateAlpha</a></code> </p>
          </td>
          <td><p>Calculate alpha</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="calculateBeta.html">calculateBeta</a></code> </p>
          </td>
          <td><p>Calculate beta</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="calculateGamma.html">calculateGamma</a></code> </p>
          </td>
          <td><p>Calculate gamma</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="calculateLogLikelihood.html">calculateLogLikelihood</a></code> </p>
          </td>
          <td><p>Calculate the log-likelihood of the data</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="calculateMeiosis.html">calculateMeiosis</a></code> </p>
          </td>
          <td><p>Estimation of Meiosis</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="calculateMissingness.html">calculateMissingness</a></code> </p>
          </td>
          <td><p>Calculate Missingness Proportions</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="calculatePopAlleleFreq.html">calculatePopAlleleFreq</a></code> </p>
          </td>
          <td><p>Calculate Allele Frequencies for SNPs from Genotype Data</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="calculateScale.html">calculateScale</a></code> </p>
          </td>
          <td><p>Calculate scale</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="calculateViterbi.html">calculateViterbi</a></code> </p>
          </td>
          <td><p>Calculate the Viterbi sequence</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="clusterSummary.html">clusterSummary</a></code> </p>
          </td>
          <td><p>A function that prints summary information of clusters</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="emissionProbDD.html">emissionProbDD</a></code> </p>
          </td>
          <td><p>The emission probabilities for 2 diploid chromosomes</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="emissionProbHD.html">emissionProbHD</a></code> </p>
          </td>
          <td><p>The emission probabilities for 1 haploid chromosome and 1 diploid chromosome</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="emissionProbHH.html">emissionProbHH</a></code> </p>
          </td>
          <td><p>The emission probabilities for 2 haploid chromosomes</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="emissionProbMissingGeno.html">emissionProbMissingGeno</a></code> </p>
          </td>
          <td><p>Calculating the emission probability sumation when missing genotype calls present</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="genotypeErrorD.html">genotypeErrorD</a></code> </p>
          </td>
          <td><p>The genotyping error probability for 1 diploid chromosome</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="genotypeErrorH.html">genotypeErrorH</a></code> </p>
          </td>
          <td><p>The genotyping error probability for 1 haploid chromosome</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="getColourPaletteMajor.html">getColourPaletteMajor</a></code> </p>
          </td>
          <td><p>IsoRelate Colour Palette for groups</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="getColourPaletteMinor.html">getColourPaletteMinor</a></code> </p>
          </td>
          <td><p><code>getColourPaletteMinor()</code> generates a specified number of shades of a given colour.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="getGenotypes.html">getGenotypes</a></code> </p>
          </td>
          <td><p>Pre-Analysis Data Processing</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="getIBDiR.html">getIBDiR</a></code> </p>
          </td>
          <td><p>Selection Significance Statistic</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="getIBDiclusters.html">getIBDiclusters</a></code> </p>
          </td>
          <td><p>Interval Cluster Networks</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="getIBDmatrix.html">getIBDmatrix</a></code> </p>
          </td>
          <td><p>Binary IBD Matrix</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="getIBDparameters.html">getIBDparameters</a></code> </p>
          </td>
          <td><p>Parameter Estimation</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="getIBDpclusters.html">getIBDpclusters</a></code> </p>
          </td>
          <td><p>Genome Cluster Networks</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="getIBDposterior.html">getIBDposterior</a></code> </p>
          </td>
          <td><p>IBD Posterior Probabilities</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="getIBDproportion.html">getIBDproportion</a></code> </p>
          </td>
          <td><p>Proportion of Pairs IBD</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="getIBDsegments.html">getIBDsegments</a></code> </p>
          </td>
          <td><p>IBD Segment Detection</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="getIBDsummary.html">getIBDsummary</a></code> </p>
          </td>
          <td><p>IBD Segment Summary</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="getOverlap.html">getOverlap</a></code> </p>
          </td>
          <td><p>function to find IBD that overlap interval</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="groupPairs.html">groupPairs</a></code> </p>
          </td>
          <td><p>Group Combinations for Analysis</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="haplotypeToGenotype.html">haplotypeToGenotype</a></code> </p>
          </td>
          <td><p>Call Genotypes from Haplotype Data</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="highlight_genes.html">highlight_genes</a></code> </p>
          </td>
          <td><p>Plasmodium Falciparum Gene Highlight Dataset</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="iRfunction.html">iRfunction</a></code> </p>
          </td>
          <td><p>Internal Function</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="isolatePairs.html">isolatePairs</a></code> </p>
          </td>
          <td><p>Pair Combinations for Analysis</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="merge_lists.html">merge_lists</a></code> </p>
          </td>
          <td><p>Internal Function</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="my_ibd.html">my_ibd</a></code> </p>
          </td>
          <td><p>IBD Segments For The Papua New Guinea Dataset</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plotIBDclusters.html">plotIBDclusters</a></code> </p>
          </td>
          <td><p>Plot Cluster Networks</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plotIBDiR.html">plotIBDiR</a></code> </p>
          </td>
          <td><p>Plot iR Statistics</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plotIBDproportions.html">plotIBDproportions</a></code> </p>
          </td>
          <td><p>Plot The Proportion of Pairs IBD</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plotIBDsegments.html">plotIBDsegments</a></code> </p>
          </td>
          <td><p>Plot IBD Segments</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="png_genotypes.html">png_genotypes</a></code> </p>
          </td>
          <td><p>Filtered Genotypes For The Papua New Guinea Dataset</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="png_ibd.html">png_ibd</a></code> </p>
          </td>
          <td><p>IBD Segments For The Papua New Guinea Dataset</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="png_parameters.html">png_parameters</a></code> </p>
          </td>
          <td><p>IBD Parameter Estimates For The Papua New Guinea Dataset</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="png_pedmap.html">png_pedmap</a></code> </p>
          </td>
          <td><p>Papua New Guinea Plasmodium Falciparum Dataset</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="roundDecimal.html">roundDecimal</a></code> </p>
          </td>
          <td><p>Round digits to specified decimal places</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="transitionProbDD.html">transitionProbDD</a></code> </p>
          </td>
          <td><p>The transition probabilities for 2 diploid chromosomes</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="transitionProbHD.html">transitionProbHD</a></code> </p>
          </td>
          <td><p>The transition probabilities for 1 haploid and 1 diploid chromosome</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="transitionProbHH.html">transitionProbHH</a></code> </p>
          </td>
          <td><p>The transition probabilities for 2 haploid chromosomes</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="trueGenotypes.html">trueGenotypes</a></code> </p>
          </td>
          <td><p>Matrices of all possible genotype combinations between pairs, given MOI</p></td>
        </tr>
      </tbody>
      </table>
    </div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#section-all-functions">All functions</a></li>
    </ul>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Lyndal Henden.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>

