<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Plot IBD Segments — plotIBDsegments • isoRelate</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  
  
<meta property="og:title" content="Plot IBD Segments — plotIBDsegments" />

<meta property="og:description" content="plotIBDsegments() plots IBD segments for pairs across the genome. IBD segments are depicted by colored blocks." />
<meta name="twitter:card" content="summary" />
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">isoRelate</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/introduction.html">isoRelate Introduction</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Plot IBD Segments</h1>
    </div>

    
    <p><code>plotIBDsegments()</code> plots IBD segments for pairs across the genome. IBD segments are depicted by colored blocks.</p>
    

    <pre class="usage"><span class='fu'>plotIBDsegments</span>(<span class='no'>ped.genotypes</span>, <span class='no'>ibd.segments</span>, <span class='kw'>interval</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>annotation.genes</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>annotation.genes.color</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>highlight.genes</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>highlight.genes.labels</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>highlight.genes.color</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>highlight.genes.alpha</span> <span class='kw'>=</span> <span class='fl'>0.1</span>,
  <span class='kw'>segment.height</span> <span class='kw'>=</span> <span class='fl'>0.5</span>, <span class='kw'>segment.color</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>number.per.page</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>fid.label</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>iid.label</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>ylabel.size</span> <span class='kw'>=</span> <span class='fl'>9</span>, <span class='kw'>add.rug</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>plot.title</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>add.legend</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>ped.genotypes</th>
      <td><p>A list containing 3 objects. See the <code>Value</code> description in <code><a href='getGenotypes.html'>getGenotypes</a></code> for more details on this input.</p></td>
    </tr>
    <tr>
      <th>ibd.segments</th>
      <td><p>A data frame containing the IBD segments inferred from pairs of isolates
See the returned <code>value</code> in <code><a href='getIBDsegments.html'>getIBDsegments</a></code> for more details.</p></td>
    </tr>
    <tr>
      <th>interval</th>
      <td><p>A vector of length 3 containing the genomic locations of a specific region to plot. This vector should contain the
chromosome ID, the start of the interval in base-pairs and the end of the interval in base-pairs; in this order respectively.
The default is <code>interval=NULL</code> which will plot the segments over all chromosomes.</p></td>
    </tr>
    <tr>
      <th>annotation.genes</th>
      <td><p>A data frame containing information on annotation genes to be included in the figure.
This data frame must have at least 5 columns of information:</p><ol>
<li><p>Chromosome (type <code>"numeric"</code> or <code>"integer"</code>)</p></li>
<li><p>Gene name (type <code>"character"</code>)</p></li>
<li><p>Start location of the gene in base-pairs (type <code>"numeric"</code> or <code>"integer"</code>)</p></li>
<li><p>End location of the gene in base-pairs (type <code>"numeric"</code> or <code>"integer"</code>)</p></li>
<li><p>Gene strand (+ or -) (type <code>"character"</code>)</p></li>
</ol><p><code>annotation.genes</code> must contain the following headers <code>chr, name, start, end</code> and <code>strand</code>.
This data frame does not have to be in a specific order, however it must contain all of the above information
with respective labels. The default is annotation.genes=NULL.</p></td>
    </tr>
    <tr>
      <th>annotation.genes.color</th>
      <td><p>A vector of characters or numeric values containing the two colors representing gene stand (positive (+) or negative (-))</p></td>
    </tr>
    <tr>
      <th>highlight.genes</th>
      <td><p>A data frame containing information of genes or regions to highlight.
The data frame must have at least 4 columns of information:</p><ol>
<li><p>Chromosome (type <code>"numeric"</code> or <code>"integer"</code>)</p></li>
<li><p>Gene name (type <code>"character"</code>)</p></li>
<li><p>Start location of the gene in base-pairs (type <code>"numeric"</code> or <code>"integer"</code>)</p></li>
<li><p>End location of the gene in base-pairs (type <code>"numeric"</code> or <code>"integer"</code>)</p></li>
</ol><p><code>highlight.genes</code> should contain the following headers <code>chr, name, start</code> and <code>end</code>.
This data frame does not have to be in a specific order, however it must contain all of the above information
with respective labels. The default is highlight.genes=NULL.</p></td>
    </tr>
    <tr>
      <th>highlight.genes.labels</th>
      <td><p>Logical. Whether to include gene names as labels in the figure. The default is <code>highlight.genes.labels=FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>highlight.genes.color</th>
      <td><p>Character string or numeric value. A single color that will be used to highlight a region/gene. The default is <code>highlight.genes.color=NULL</code>.</p></td>
    </tr>
    <tr>
      <th>highlight.genes.alpha</th>
      <td><p>Numeric. A single value between 0 and 1 indicating the gene color transparency. The default is <code>highlight.genes.alpha=0.1</code>.</p></td>
    </tr>
    <tr>
      <th>segment.height</th>
      <td><p>A numeric value giving the hight of IBD segment blocks, such that 0 &lt; segment.height &lt;= 1. The default is <code>segment.height=0.5</code>.</p></td>
    </tr>
    <tr>
      <th>segment.color</th>
      <td><p>A vector of characters or numeric values denoting the color of the segments to be plotted.
Two colors must be specified, one for segments with 1 allele IBD and one for segments with 2 alleles IBD.</p></td>
    </tr>
    <tr>
      <th>number.per.page</th>
      <td><p>A numeric value indicating the maximum number of IBD pairs to plot in a single graphics window. The default is
<code>number.per.page=NULL</code> which will plot all IBD pairs in a single window. This may not be ideal when there are many IBD pairs. If
<code>number.per.page</code> is set, it is recommended to plot the output to a file as opposed to the R plotting window.</p></td>
    </tr>
    <tr>
      <th>fid.label</th>
      <td><p>Logical. If <code>fid.label=TRUE</code>, family IDs will be included in the y-axis labels; otherwise family IDs will be omitted.
The default is <code>add.fid.name=TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>iid.label</th>
      <td><p>Logical. If <code>iid.label=TRUE</code>, isolate IDs will be included in the y-axis labels; otherwise isolate IDs will be omitted.
The default is <code>add.iid.name=TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>ylabel.size</th>
      <td><p>A numeric value indicating the size of the y-axis labels if drawn. The default is <code>ylabel.size=9</code>.</p></td>
    </tr>
    <tr>
      <th>add.rug</th>
      <td><p>Logical. Whether to include SNP positions as a rug in the figure. The default is <code>add.rug=FALSE</code></p></td>
    </tr>
    <tr>
      <th>plot.title</th>
      <td><p>A character string of a title to be added to the figure The default is <code>plot.title=NULL</code> which does not add a title to the plot.</p></td>
    </tr>
    <tr>
      <th>add.legend</th>
      <td><p>Logical. If <code>add.legend=TRUE</code>, a legend specifying the IBD status (1 allele IBD or 2 alleles IBD) will be included. The default is <code>add.legend=TRUE</code>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <p><code><a href='getGenotypes.html'>getGenotypes</a></code> and <code><a href='getIBDsegments.html'>getIBDsegments</a></code>.</p>
    

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#see-also">See also</a></li>
          </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Lyndal Henden.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>

